### ISSUE TITLE
About the loss function design

### ISSUE THREAD

**Issue by HuangChiEn** [ZhichengHuang/CMAE#5]
Hello, thanks for releasing this amazing work.

I have some questions about the loss function design. The loss function could be the BYOL style loss as well as the contrastive loss. To simplify the implementation, we choice to apply BYOL style loss, but not sure that: "is it only calculate the asymmetric loss and backpropagate to the network"? 

For example
```
class CMAE :
    def __init__ ( self, ... ) :  # omit args
        self.online_enc, self.target_enc = ..., ... # omit declaration 
        self.pixl_dec, self.feat_dec = ..., ... # omit declaration
        # BYOL-style proj-pred struct
        self.proj, self.pred, self.momentum_proj = ..., ..., ... # omit declaration
 
    def forward ( self, X ) :
        v_onl, v_tar = X
        # omit masking..
        # Suppose target_encoder forward implement the mean-pooling
        onl_p, tar_feat = self.online_enc(v_onl), self.target_enc(v_tar)
        im_p, onl_feat = self.pixl_dec(onl_p), self.feat_dec(onl_p)

        # predicted representation and projected representation
        p, z = self.pred( self.proj(onl_feat) ), self.momentum_proj(tar_feat)
        # omit BYOL loss implement..
        loss = BYOL_loss(p, z) 
        
        # No symmetric term ?  
        # ? p, z = self.pred( self.proj(tar_feat) ), self.momentum_proj(onl_feat)
```

Any suggestion will be appreciated !!

**Comment by ZhichengHuang** [comment]
We did not validate the effect of small batch size, but in general, large batch size can improve the performance of contrastive learning.  The following is the loss function:
```python
    self.criterion = nn.CrossEntropyLoss()



    def forward(self,x,pred_pixel,mask_s,proj_s,proj_t):
        target = self.patchify(x)
        with torch.no_grad():
            mean = target.mean(dim=-1, keepdim=True)
            var = target.var(dim=-1, keepdim=True)
            target = (target - mean) / (var + 1.e-6) ** .5

        loss_rec = (pred_pixel - target) ** 2
        loss_rec = loss_rec.mean(dim=-1)
        loss_rc = (loss_rec * mask_s).sum() / mask_s.sum()

        pred_s = self.predictor(proj_s)

        pred_s = F.normalize(pred_s.squeeze(dim=1),dim=1,p=2)
        proj_t = F.normalize(proj_t.squeeze(dim=1),dim=1,p=2)

        proj_t = concat_all_gather(proj_t)

        score = torch.matmul(pred_s,proj_t.transpose(1, 0).detach())

        score = score / self.t

        bs = score.size(0)
        label = (torch.arange(bs, dtype=torch.long) +
                 bs * torch.distributed.get_rank()).cuda()

        losses = dict()
        losses['loss_ct'] =  2 * self.t * self.criterion(score, label)
        losses['loss_rc'] =  loss_rc

        return losses
```

**Comment by HuangChiEn** [comment]
> ```python
> self.criterion(score, label)
> ```

This code snippet seems the implementation of SimCLR loss, do you have the BYOL style version ?

**Comment by HuangChiEn** [comment]
> We did not validate the effect of small batch size, but in general, large batch size can improve the performance of contrastive learning. The following is the loss function:
> 
> ```python
>     self.criterion = nn.CrossEntropyLoss()
> 
> 
> 
>     def forward(self,x,pred_pixel,mask_s,proj_s,proj_t):
>         target = self.patchify(x)
>         with torch.no_grad():
>             mean = target.mean(dim=-1, keepdim=True)
>             var = target.var(dim=-1, keepdim=True)
>             target = (target - mean) / (var + 1.e-6) ** .5
> 
>         loss_rec = (pred_pixel - target) ** 2
>         loss_rec = loss_rec.mean(dim=-1)
>         loss_rc = (loss_rec * mask_s).sum() / mask_s.sum()
> 
>         pred_s = self.predictor(proj_s)
> 
>         pred_s = F.normalize(pred_s.squeeze(dim=1),dim=1,p=2)
>         proj_t = F.normalize(proj_t.squeeze(dim=1),dim=1,p=2)
> 
>         proj_t = concat_all_gather(proj_t)
> 
>         score = torch.matmul(pred_s,proj_t.transpose(1, 0).detach())
> 
>         score = score / self.t
> 
>         bs = score.size(0)
>         label = (torch.arange(bs, dtype=torch.long) +
>                  bs * torch.distributed.get_rank()).cuda()
> 
>         losses = dict()
>         losses['loss_ct'] =  2 * self.t * self.criterion(score, label)
>         losses['loss_rc'] =  loss_rc
> 
>         return losses
> ```

I see, thanks for your information!!
